subsys/net/lib/http/http_server.c

```diff
@@ -0,0 +1,1235 @@
+{
+	ctx_client->server_state = HTTP_SERVER_FRAME_WINDOW_UPDATE_STATE;
+
+	return 0;
+}
+
+int enter_http_frame_priority_state(struct http_client_ctx *ctx_client)
+{
+	ctx_client->server_state = HTTP_SERVER_FRAME_PRIORITY_STATE;
+
+	return 0;
+}
+
+int enter_http_frame_rst_stream_state(struct http_server_ctx *ctx_server,
+				      struct http_client_ctx *ctx_client, int client_index)
+{
+	ctx_client->server_state = HTTP_SERVER_FRAME_RST_STREAM_STATE;
+
+	return 0;
+}
+
+int enter_http_frame_goaway_state(struct http_server_ctx *ctx_server,
+				  struct http_client_ctx *ctx_client, int client_index)
+{
+	ctx_client->server_state = HTTP_SERVER_FRAME_GOAWAY_STATE;
+
+	return 0;
+}
+
+int enter_http_http_done_state(struct http_server_ctx *ctx_server,
+			       struct http_client_ctx *ctx_client, int client_index)
+{
+	ctx_client->server_state = HTTP_SERVER_DONE_STATE;
+
+	return 0;
+}
+
+int handle_http_preface(struct http_client_ctx *ctx_client)
+{
+	LOG_DBG("HTTP_SERVER_PREFACE_STATE.");
+	if (ctx_client->offset < strlen(preface)) {
+		/* We don't have full preface yet, get more data. */
+		return -EAGAIN;
+	}
+
+	if (strncmp(ctx_client->buffer, preface, strlen(preface)) != 0) {
+		ctx_client->server_state = HTTP_SERVER_REQUEST_STATE;
+	} else {
+		ctx_client->server_state = HTTP_SERVER_FRAME_HEADER_STATE;
+
+		ctx_client->offset -= strlen(preface);
+		memmove(ctx_client->buffer, ctx_client->buffer + strlen(preface),
+			ctx_client->offset);
+	}
+
+	return 0;
+}
+
+int handle_http1_request(struct http_server_ctx *ctx_server, struct http_client_ctx *ctx_client,
+			 int client_index)
+{
+	LOG_DBG("HTTP_SERVER_REQUEST.");
+	const char *data;
+	int len;
+	int total_received = 0;
+	int offset = 0;
+
+	http_parser_init(&parser, HTTP_REQUEST);
+	http_parser_settings_init(&parserSettings);
+	parserSettings.on_header_field = on_header_field;
+	parserSettings.on_url = on_url;
+	http_parser_execute(&parser, &parserSettings, ctx_client->buffer + offset,
+			    ctx_client->offset);
+
+	total_received += ctx_client->offset;
+	offset += ctx_client->offset;
+
+	if (offset >= HTTP_SERVER_MAX_RESPONSE_SIZE) {
+		offset = 0;
+	}
+
+	if (has_upgrade_header == false) {
+		const char *response = "HTTP/1.1 101 Switching Protocols\r\n"
+				       "Connection: Upgrade\r\n"
+				       "Upgrade: h2c\r\n"
+				       "\r\n";
+		if (sendall(ctx_client->client_fd, response, strlen(response)) < 0) {
+			close_client_connection(ctx_server, client_index);
+		}
+
+		memset(ctx_client->buffer, 0, sizeof(ctx_client->buffer));
+		ctx_client->offset = 0;
+		ctx_client->server_state = HTTP_SERVER_PREFACE_STATE;
+
+	} else {
+		data = content_200;
+		len = sizeof(content_200);
+
+		enum http_method method = parser.method;
+		const char *method_str = http_method_str(method);
+
+		LOG_DBG("HTTP Method: %s", method_str);
+
+		if (strncmp(method_str, "GET", 3) == 0 && strncmp(url_buffer, "/results", 8) == 0) {
```

Comment:
Why is a test-specific resource ("/result") hard-coded in the generic http server library code?

The http_server in it's present state very much lacks a **CRITICAL** piece of functionality - i.e. "routing" a request to a handler based on the name and other application-specific properties of the resource.

Resources must be defined using `HTTP_RESOURCE_DEFINE()`. This will place all http resources into [Iterable Sections](https://docs.zephyrproject.org/latest/kernel/iterable_sections/index.html).

That is the only way that the server can reasonably expect to iterate over a collection (array) of resources that are declared in separate compilation units.

There are already test cases demonstrating how that is done.

The idea is to iterate over each resource associated with a particular service (defined by `HTTP_SERVICE_DEFINE()`). The resource "path" can then be compared with the http request path, etc...